src/main/java/updater/DownloadCountingOutputStream.java

```diff
@@ -0,0 +1,51 @@
+package updater;
+
+import javafx.application.Platform;
+import javafx.beans.property.DoubleProperty;
+import javafx.beans.property.SimpleDoubleProperty;
+import javafx.beans.value.ObservableValue;
+import org.apache.commons.io.output.CountingOutputStream;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.concurrent.Callable;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * This class counts the bytes that have been downloaded and tell its listener(s) about current percentage downloaded
+ */
+public class DownloadCountingOutputStream extends CountingOutputStream {
+    private final long totalFileSizeInBytes;
+    private final DoubleProperty percentageDownloaded; // 0.0 to 1.0
+
+    /**
+     *
+     * @param out output stream
+     * @param totalFileSizeInBytes total file size to be downloaded
+     */
+    public DownloadCountingOutputStream(OutputStream out, long totalFileSizeInBytes) {
+        super(out);
+
+        this.totalFileSizeInBytes = totalFileSizeInBytes;
+        percentageDownloaded = new SimpleDoubleProperty(0);
+    }
+
+    public void addMethodThatListen(Consumer<Double> method) {
+        percentageDownloaded.addListener(
+                (ObservableValue<? extends Number> observable, Number oldValue, Number newValue) ->
+                        Platform.runLater(() -> method.accept(newValue.doubleValue())));
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Update percentage downloaded
+     */
+    @Override
+    protected void afterWrite(int n) throws IOException {
+        super.afterWrite(n);
+
+        percentageDownloaded.set(1.0 * this.getByteCount() / this.totalFileSizeInBytes);
```

Comment:
I'm not sure this is the right place for Platform.runLater. Without context, we won't know why it's needed, because this class could be used to update the progress of anything, not necessarily JavaFX nodes. We can put it in at the point where we actually modify the UI maybe?